### PR TITLE
Convert names tab to Preact/Tailwind

### PR DESCRIPTION
**Pull request by enellis**
This is still very much work in progress. I'm opening this PR now to have a suitable place for discussion.

Edit: I will be focusing on converting the metadata first, as it is rendered in the names tab in some cases (https://github.com/birchill/10ten-ja-reader/pull/2114#discussion_r1854962435).

**Todo:**
- [x] #2118 
- [x] `tag-` classes
- [x] Remove tests in popups.html
- [x] Correct selected styles
- [x] Check for unused CSS

### CODE REVIEW

File: src/content/popup/names.ts [hunk 1]
```diff
@@ -22,143 +17,26 @@ export function renderNamesEntries({
   more: boolean;
   options: ShowPopupOptions;
 }): HTMLElement {
-  const namesTable = html('div', { class: 'name-table entry-data' });
+  const containerElement = html('div', {
+    /* Make sure it's possible to scroll all the way to the bottom of each kanji
+     * table. */
+    style:
+      'scroll-snap-align: start; scroll-margin-bottom: var(--expand-button-allowance);',
+  });
```
Comment (enellis) [inline src/content/popup/names.ts:25]: Copied from the kanji tab. Is this also needed in the names tab?

File: src/content/popup/names.ts [hunk 2]
```diff
@@ -22,143 +17,26 @@ export function renderNamesEntries({
...
-      isCombinedResult: true,
-      matchLen,
+      copyState: options.copyState,
+      onStartCopy: options.onStartCopy,
       meta: options.meta,
-    });
-    if (metadata) {
-      namesTable.append(metadata);
-    }
-  }
```
Comment (enellis) [inline src/content/popup/names.ts:38]: When do we show metadata in the names tab? Can't think of a scenario off the top of my head.

File: src/content/popup/tailwind.config.js [hunk 3]
```diff
@@ -46,6 +50,14 @@ export default {
...
           { lineHeight: 'calc(1.5 * var(--base-font-size))' },
         ],
+        kana: [
+          'calc(18 / 14 * var(--base-font-size))',
+          { lineHeight: 'calc(1.7 * var(--base-font-size))' },
+        ],
+        kanji: [
+          'calc(20 / 14 * var(--base-font-size))',
+          { lineHeight: 'calc(1.8 * var(--base-font-size))' },
+        ],
```
Comment (enellis) [inline src/content/popup/tailwind.config.js:60]: How should these be named? Semantically, as they are currently, or according to the common tailwind sizes?

File: src/content/popup/names.ts (hunk 1 above)
Comment (birtles) [reply src/content/popup/names.ts]: Good question. I'd have to experiment I'm afraid. If it's not currently used in the names tab then I think it's safe to leave it off for now.

Comment (enellis) [reply src/content/popup/names.ts]: Yeah, scroll snap is currently not used in the names tab at all, so there should be no reason to add this style.

File: src/content/popup/tailwind.config.js (hunk 3 above)
Comment (enellis) [reply src/content/popup/tailwind.config.js]: I think it's best to name them based on Tailwind's size conventions, though I'm not entirely sure of the exact mappings.
If 14 is 'base,' then would `15 / 14` be lg, `16 / 14` xl, `17 / 14` 2xl ...?

Comment (birtles) [reply src/content/popup/tailwind.config.js]: Yes, I agree. I wrote a few thoughts about this [over in the metadata PR](https://github.com/birchill/10ten-ja-reader/pull/2118#discussion_r1858306944).

File: src/content/popup/Names/NameEntryTW.tsx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        // };
+
+        return (
+          tagText && (
+            <span class="w-def">
+              <span
+                class={classes(
+                  'tp-rounded tp-text-2xs tp-px-1 tp-whitespace-nowrap',
+                  '!tp-border-solid !tp-border !tp-border-[--tag-border]',
+                  `tag-${tag}`
```
[the remark this answers is not in the document]
Comment (enellis) [reply src/content/popup/Names/NameEntryTW.tsx]: Thanks for the link!
Unfortunately, I couldn't think of a very elegant solution yet. For now, I added the CSS variables `--tag-[green/pink/blue]-text` and `--tag-[green/pink/blue]-border` for every theme and apply them accordingly in the `Tag` component in `NameEntry.tsx`. The fact that it's not possible to use JS variables in Tailwind classes makes it somewhat verbose, though. Could you have a look at the `add Tailwind styles for NameEntry component` commit? Maybe you have an idea on how to make the code a bit cleaner.

[the remark this answers is not in the document]
Comment (birtles) [reply src/content/popup/Names/NameEntryTW.tsx]: I think what you have looks great. I'd probably add `-bg` to the background color CSS variables. Also I think by using CSS variable fallback values we could possibly make it a bit simpler like this: https://gist.github.com/birtles/7c33984f6b843f6af1a6975f29291163

[the remark this answers is not in the document]
Comment (enellis) [reply src/content/popup/Names/NameEntryTW.tsx]: Looks very good, thank you!

**Review by birtles — APPROVED** [review]
Amazing work, thank you so much!

In testing this, however, I noticed that at some point in the past we seemed to have regressed the behavior when looking up 64,040円. I'm pretty sure in the past when we looked that up we'd only show the names tab (and maybe the kanji tab too) but now we also show the words tab with only the metadata.

When I get chance, I'll try to run `git bisect` to see where that regressed.

File: src/content/popup/Names/NameEntry.tsx [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+              'hh:tp-ease-out',
+              'hh:hover:tp-transition-none',
+            ]
+          : [])
+      )}
+      onPointerUp={props.onPointerUp}
+      onClick={props.onClick}
+    >
+      <div class="tp-space-x-4" lang="ja">
+        {props.entry.k && (
```
[the remark this answers is not in the document]
Comment (enellis) [reply src/content/popup/Names/NameEntry.tsx]: Yes, good call!

File: src/content/popup/Names/NameTable.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    <div
+      class={classes('tp-my-2', entries.length > 4 ? 'tp-columns-2' : '')}
+      ref={namesTable}
+    >
+      <div style={{ 'column-span': 'all' }}>
+        {props.meta && (
+          <MetadataContainer
+            fxData={props.fxData}
+            preferredUnits={props.preferredUnits}
+            isCombinedResult={true}
```
[the remark this answers is not in the document]
Comment (enellis) [reply src/content/popup/Names/NameTable.tsx]: Good to know :)

File: src/content/popup/Names/NameEntry.tsx [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+              'hh:tp-transition-colors hh:interactive:tp-duration-100',
+              'hh:tp-ease-out',
+              'hh:hover:tp-transition-none',
+            ]
+          : [])
+      )}
+      onPointerUp={props.onPointerUp}
+      onClick={props.onClick}
+    >
+      <div class="tp-space-x-4" lang="ja">
```
[the remark this answers is not in the document]
Comment (enellis) [reply src/content/popup/Names/NameEntry.tsx:53]: Okay, so it's okay to leave it as it is here? I will definitely keep this option in mind for the next time :)

**Comment by enellis** [comment]
> When I get chance, I'll try to run `git bisect` to see where that regressed.

Did it real quick, with the result that ec7d6db23d51cca11fdd555be9dde0f937e72242 is the first bad commit.

File: src/content/popup/Names/NameEntry.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (birtles) [reply src/content/popup/Names/NameEntry.tsx:53]: Yes, that's fine. Thanks!

**Comment by birtles** [comment]
> > When I get chance, I'll try to run `git bisect` to see where that regressed.
> 
> Did it real quick, with the result that [ec7d6db](https://github.com/birchill/10ten-ja-reader/commit/ec7d6db23d51cca11fdd555be9dde0f937e72242) is the first bad commit.

Thank you!

**Pull request merged**